### ISSUE TITLE
Mapping ALT/META (:map <m-k> blah) keys get scrambled.

### ISSUE THREAD

**Issue by AmokHuginnsson** [vim/vim#2588]
### Issue
Any mapping that uses `<m-key>` or `<a-key>` gets scrambled into some other code.
The same behavior can be observer in all terminal configurations tested: (xterm, rxvt, Eterm, Linux console, tmux/no tmux, screen/no screen).

### Steps to reproduce:
`:map <m-s> blah|map`

### Expected result
`<m-s> blah`

### Actual result
`ó blah`

### Environment
- VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Dec 21 2016 20:05:08)
- Ubuntu 18.04
- terminator 1.91
- tmux 2.6
- `echo $TERM` = `screen-256color-bce`
- `set term?` = `term=screen-256color-bce`

**Comment by chrisbra** [comment]
Originally, in the old ASCII days, the alt key worked by flipping the high bit of the key pressed and that is how Vim in a terminal sees the input. That is also the reason, that pressing `ó` will also trigger your map.
That is basically an implementation detail, that is visible to the user and therefore expected.

Closing.

**Comment by AmokHuginnsson** [comment]
But this is completely wrong!  
Why pressing LALT+s would have the same effect as pressing RALT+o?  
You want to tell me that vim does not distinguish between ALT and ALTgr?  
I want to have LALT+s mapping and still be able to type ó.  
I know I can map ^[s, but it has other drawbacks.  
Are there any plans to fix this?

**Comment by chrisbra** [comment]
Note, this is how input handling was done (https://en.wikipedia.org/wiki/Alt_key)
The basic input system is complex and not easily changeable without introducing bugs, note that in the terminal Vim has no way to receive the key itsself, it must get it from the terminal. However there have been changes proposed like here:
https://groups.google.com/d/msg/vim_dev/Ym6D-kWIsyo/-XOw7E8LbN0J
however it is not clear this won't break on some obscure systems or under some circumstances.

It might work fine in the GUI, Vim can distinguish more keys there.

**Comment by tonymec** [comment]
@AmokHuginnsson : It is not that Vim doesn't distinguish between left and right Alt, it is that when running in a terminal, Vim gets its keys only after the terminal (and, in Linux, the X server) have processed them. There are keys there which Vim simply doesn't get because the terminal or the window manager preempt them for their own uses, and there are key combinations which Vim cannot tell apart, simply because they result in the same byte sequences being passed to it.

As Christian says, gvim has more control over the interpretation of keys and key combinations, because there is one fewer middleman: there is no external terminal here to interfere with the communication between Vim and the keyboard. On X11 systems, the X11 layer is still there but it interprets the keystrokes the same way for Vim as for any other GUI: on my system, for instance, AltGr+o doesn't give ó but œ, while Alt+s gives me (in gvim) the Syntax menu: they are clearly different. While typing in this textarea in my browser, AltGr+o still gives œ, while Alt+s does nothing because there is no menu whose name starts with S.

**Comment by AmokHuginnsson** [comment]
@chrisbra 
Thread you cited is 7 years old, if nothing was done with this, then the idea died in 2010. I understand that there is not enough pressure to fix this.  
Still I consider this a bug in vim, and by 'this' I mean conflating LALT+s (which rest of my system correctly recognizes as ALT+s) with ó (obtained in my country by RALT+o).  
I would still prefer that some bug report would be opened for this issue, if my current wording was not precise enough I can open new issue and add this conversation as part of a bug description.

@tonymec
I know for a fact that there are editors working in Linux terminal that have no problem with behaving correctly in this regard, it means that this bug can be fixed.

**Comment by mgedmin** [comment]
I believe NeoVim correctly recognizes `<esc>x` sequences many modern terminals issue for Alt-x as `<M-x>`.  Vim continues to assume the older x-but-with-hi-bit-set convention out of some sort of backwards compatibility sentiment I don't quite understand myself.

If you continue to use Vim rather than NeoVim, you can explicitly map `<esc>x` instead of `<M-x>`.

**Comment by AmokHuginnsson** [comment]
@mgedmin 
It is true, I submitted this bug report because I found out that NeoVim can handle `<M-key>` properly. I did not want to invoke it's name, to not to stir, but cat is out of the bag. All explanations given here as of why this bug cannot be fixed really break my heart. I have been faithful vimmer for last 17 years. I want to stick with vim. Right now my vimrc is plastered with whole bunch of `if has( "nvim" )`.

**Comment by brammool** [comment]
Marius Gedminas wrote:

> I believe NeoVim correctly recognizes `<esc>x` sequences many modern
> terminals issue for Alt-x as `<M-x>`.  Vim continues to assume the
> older x-but-with-hi-bit-set convention out of some sort of backwards
> compatibility sentiment I don't quite understand myself.

I strongly discourage having the terminal send <Esc>x for Alt-X.  It
will get you in trouble sooner or later, because it's impossible to
distinguish from pressing the <Esc> key and then the x key.  Only some
tricky timeout might work, and that's bound to break, e.g. if your
system is busy with something.

Xterm can send modifier flags with a key, that's a much more reliable
method.  But it's tricky to decode.

> If you continue to use Vim rather than NeoVim, you can explicitly map
> `<esc>x` instead of `<M-x>`.

-- 
GOD: That is your purpose Arthur ... the Quest for the Holy Grail ...
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mgedmin** [comment]
> I strongly discourage having the terminal send <Esc>x for Alt-X.  It
will get you in trouble sooner or later, because it's impossible to
distinguish from pressing the <Esc> key and then the x key. 

It's very useful to emulate Alt-keystrokes by pressing <Esc> followed by the key when you're working with a keyboard that lacks an Alt key (e.g. an ssh app on my Android phone).

Meanwhile the 8th bit convention makes it impossible to distinguish Alt-X from a user typing ø on their national keyboard.  (It's why I had to stop using Alt-mappings in insert mode in gvim back in 2000, because I could then no longer type text in Lithuanian!)

**Comment by AmokHuginnsson** [comment]
@mgedmin 
To be perfectly clear, no one is suggesting to remove `<esc>x` emulation.  
What is being suggested it is to allow vim to understand physical ALT+x press and its symbolic representation `<m-x>` for what they really are.

**Comment by dylnmc** [comment]
@AmokHuginnsson 

After some careful testing, it appears as though this will fix most meta key mappings in terminal vim. Although, I could not get `<m-@>` to work in any terminal other than xterm, `<m-[>` and `<m-]>` didn't work in any terminal, `<m-O>` did not seem to cooperate in most terminals except for xterm (that's capital "Oh"), `<c-up>` and `<c-down>` did not work in any terminal, and a few other nuances. However, mapping pretty much anything works in gvim. I didn't test neovim, but I heard mapping there works pretty well.

```vim
if !has('nvim') && !has('gui_running')
  set ttimeoutlen=5
  " set up Meta to work properly for most keys in terminal vim
  " NOTE: these do not work: <m-space>,<m->>,<m-[>,<m-]>,<m-{up,down,left,right}>
  " NOTE: <m-@>,<m-O> only work in xterm and gvim - not st, urxvt, etc
  " NOTE: map <m-\|> or <m-bar>
  for ord in range(33,61)+range(63,90)+range(92,126)
    let char = ord is 34 ? '\"' : ord is 124 ? '\|' : nr2char(ord)
    exec printf("set <m-%s>=\<esc>%s", char, char)
    if exists(':tnoremap') " fix terminal control sequences
      exec printf("tnoremap <silent> <m-%s> <esc>%s", char, char)
    endif
  endfor
  " set up <c-left> and <c-right> properly
  " NOTE: if below don't work, compare with ctrl-v + CTRL-{LEFT,RIGHT} in INSERT mode
  " NOTE: <c-up>,<c-down> do not work in any terminal
  exe "set <c-right>=\<esc>[1;5C"
  exe "set <c-left>=\<esc>[1;5D"
endif
```

The `tnoremap` should fix a small bug in the terminal as well. Overall, this has treated me quite well, and I have had no major issues in vim 8.1 and 8.2.

Here is how you can test it:

```vim
sil update
let [s:lhs,s:rhs] = [{'|':'bar'}, {'|':'bar', "'":"''"}]
for ord in range(33,61)+range(63,90)+range(92,126)
    let char = nr2char(ord)
    let [l,r] = [get(s:lhs, char, char), get(s:rhs, char, char)]
    exe printf("nnore <m-%s> :echo 'M-%s'<cr>", l, r)
    exe printf("nnore <f20><m-%s> :let g:lastmap = '%s'<cr>", l, r)
endfor
for ord in range(33,61)+range(63,90)+range(92,126)
    let char = nr2char(ord)
    sil call feedkeys("\<f20>\<esc>".char, 'Tx')
    if get(g:, 'lastmap') != get(s:lhs, char, char)
        echom 'Broken' '<m-'.char.'>'
    endif
endfor
if &mod | sil earlier 1f | endif
```

For me it, it says the following:

- `Broken <m-@>` - right; `<m-@>` is broken
- `Broken <m-O>` - right; `<m-O>` is broken for some reason

Additionally, you can manually test the keys by issuing command sequences with alt. For example, <kbd>Alt</kbd><kbd>#</kbd> should print `M-#`.

So you don't have to test manually or if you get lost, here is a gif (in `st`):

![foo](https://user-images.githubusercontent.com/7635158/94093694-d885a300-fdeb-11ea-8593-2276f743ab64.gif)
